RavenDB-17793 : Prefixed Sharding

aviv86 opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-17793/Shard-by-prefix

### Additional description

- Completed the work started here https://github.com/ravendb/ravendb/pull/15051
- Prefixed sharding gives you the option to control documents distribution between shards based on the document-id prefix
- Users can configure PrefixedSharding settings in database creation (via db-record), or at a later stage by using
dedicated operations (Add, Delete, Update). 
You cannot modify PrefixedSharding settings via db-record after database creation

```
// configure prefixed sharding settings during database creation
await store.Maintenance.Server.SendAsync(new CreateDatabaseOperation(new DatabaseRecord
{
    DatabaseName = "nw",
    Sharding = new ShardingConfiguration
    {
        Shards = new Dictionary<int, DatabaseTopology>
        {
            [0] = new DatabaseTopology(), 
            [1] = new DatabaseTopology(), 
            [2] = new DatabaseTopology()
        },
        Prefixed =
        [
            new PrefixedShardingSetting
            {
                Prefix = "eu/", 
                Shards = [0, 1]
            },
            new PrefixedShardingSetting
            {
                Prefix = "usa/", 
                Shards = [1, 2]
            }
        ]
    }
}));

// add prefixed setting
await store.Maintenance.SendAsync(new AddPrefixedShardingSettingOperation(new PrefixedShardingSetting
{
    Prefix = "asia/",
    Shards = [1, 2]
}));

// delete prefixed setting
await store.Maintenance.SendAsync(new DeletePrefixedShardingSettingOperation("eu/"));

// update shards distribution for a specific prefixed setting
await store.Maintenance.SendAsync(new UpdatePrefixedShardingSettingOperation(new PrefixedShardingSetting
{
    Prefix = "usa/",
    Shards = [0, 2]
}));


```


- Prefixes are sorted in a descending order - so that we can easily support having a prefix of prefix (e.g. `users/` and `users/usa/`)
- Can have up to 4K prefixes
- Each prefix setting gets a range of 1M buckets. buckets 0 - 1M are used for "regular" (non-prefixed) items.
e.g. :

```
// docs starting with 'eu/' will reside on shard #0
// docs starting with 'asia/' will reside on shard #1
// docs starting with 'us/' will reside on shard #0 or shard #2

DbRecord.Sharding.Prefixed = 
{ 
	'eu/' 	: [shard 0], 
	'asia/' : [shard 1], 
	'us/' 	: [shard 0, shard 2] 	
} 

DbRecord.Sharding.BucketRanges = 

// default bucket ranges
shard 0 : [0, 333K] 
shard 1 : [333K, 666K]
shard 2 : [666K, 1M] 

// prefixed ranges
shard 0 : [1M, 2M] // 'eu/' range
shard 1 : [2M, 3M]  // 'asia/' range
shard 0 : [3M, 3.5M]  // 'us/' range
shard 2 : [3.5M, 4M]  // 'us/' range

```

Some restrictions : 
- When adding or deleting a prefixed sharding setting, you cannot have any documents in the database the start with this prefix.
we validate that on the server side, but in a race condition some docs can slip between the time of validation and command execution - which can lead to mapping problem (ids mapped to *wrong* buckets). It is therefore the user's responsibly to make sure no matching docs exist when adding/deleting a prefix.
- Update - cannot remove shards from prefixed sharding setting if they still have buckets assigned to them (for this prefix).
in order to do that, first you'll need to move all the shard's buckets to a different shard
- Adding a shard to an existing prefix setting distribution will not assign any buckets for the new shard. Users need to manually 
move buckets to the new shard.
- Cannot remove the last copy of a shard from the database if this shard is still a part of any PrefixedSharding settings
- Prefixes get precedence over anchoring when choosing shard number for an id, e.g. :
```
* dbRecord.Sharding.Prefixed = { "products/" : [shard #2] }
* id = "companies/1" => bucket = 12345 => belongs to shard #0 
* id = "products/1$companies/1" => bucket is 12345 + 1_000_000 (RangeStart of `products/` prefix) =>
 belongs to shard #2
```

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [X] New feature

### How risky is the change?

- [ ] Low 
- [X] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [X] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [X] No

### Documentation update

- [x] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [ ] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [x] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [x] It has been verified by manual testing

### Testing by RavenDB QA team

- [x] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [ ] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [x] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [ ] No UI work is needed

karmeli87 commented on src/Raven.Server/Documents/Handlers/Admin/AdminShardingHandler.cs:
```diff
@@ -22,6 +23,24 @@ public async Task ExecuteMoveDocuments()
             HttpContext.Response.StatusCode = (int)HttpStatusCode.NoContent;
         }
 
+        [RavenAction("/databases/*/admin/sharding/prefixed/add", "POST", AuthorizationStatus.DatabaseAdmin)]
+        public Task AddPrefixedShardingSetting()
+        {
+            throw new NotSupportedInShardingException("This operation is not available from a specific shard");
+        }
+
+        [RavenAction("/databases/*/admin/sharding/prefixed/delete", "DELETE", AuthorizationStatus.DatabaseAdmin)]
+        public Task DeletePrefixedShardingSetting()
+        {
+            throw new NotSupportedInShardingException("This operation is not available from a specific shard");
+        }
+
+        [RavenAction("/databases/*/admin/sharding/prefixed/update", "POST", AuthorizationStatus.DatabaseAdmin)]
+        public Task UpdatePrefixedShardingSetting()
+        {
+            throw new NotSupportedInShardingException("This operation is not available from a specific shard");
+        }
+
```
wonder if we should move all this to `ShardedAdminDatabaseHandler`

ppekrol replied:
why? isn't this a better handler or I'm missing sth?

karmeli87 commented on src/Raven.Server/Documents/Sharding/Handlers/Admin/ShardedAdminShardingHandler.cs:
```diff
@@ -13,5 +27,177 @@ public async Task ExecuteMoveDocuments()
+                var clusterTopology = ServerStore.GetClusterTopology(context);
+                var urls = shardingConfiguration.Orchestrator.Topology.Members.Select(clusterTopology.GetUrlFromTag).ToArray();
+
+                if (await AssertNoDocumentsStartingWith(context, setting.Prefix, urls) == false)
```
this is a remote call under read tx, can we scope it only to the `GetClusterTopology`?

aviv86 replied:
fixed

karmeli87 commented on src/Raven.Server/Documents/Sharding/Handlers/Admin/ShardedAdminShardingHandler.cs:
```diff
@@ -13,5 +27,177 @@ public async Task ExecuteMoveDocuments()
+        {
+            using (var requestExecutor = RequestExecutor.CreateForServer(urls, database ?? DatabaseName, ServerStore.Server.Certificate.Certificate, DocumentConventions.DefaultForServer))
+            {
+                var command = new GetDocumentsCommand(requestExecutor.Conventions, startWith: prefix,
```
this doesn't cover the case if the shard has replication factor > 1, but I guess it also part of the docs 
>It is therefore the user's responsibly to make sure no matching docs exist when adding/deleting a prefix.

karmeli87 commented on src/Raven.Server/Documents/Sharding/Handlers/Admin/ShardedAdminShardingHandler.cs:
```diff
@@ -13,5 +27,177 @@ public async Task ExecuteMoveDocuments()
+        }
+
+        private static void AssertValidShardsDistribution(PrefixedShardingSetting oldSetting, PrefixedShardingSetting updatedSetting, ShardingConfiguration configuration)
+        {
```
If I get this right.. would it be simpler to have 2 different commands of `AddShardToPrefix` & `DeleteShardFromPrefix` ?

aviv86 replied:
it would make this part of the code a bit simpler, yes
but for the users I think it will be nicer and simpler to have a single command to handle both

karmeli87 reviewed (commented):
In general, LGTM

ppekrol commented on src/Raven.Server/ServerWide/Commands/Sharding/UpdatePrefixedShardingSettingCommand.cs:
```diff
@@ -0,0 +1,36 @@
+        public override void UpdateDatabaseRecord(DatabaseRecord record, long etag)
+        {
+            var location = record.Sharding.Prefixed.BinarySearch(Setting, PrefixedSettingComparer.Instance);
+            Debug.Assert(location >= 0, $"Prefixed setting '{Setting.Prefix}' was not found in sharding configuration");
```
in release we will get index out of bounds ex?

aviv86 replied:
fixed

ppekrol commented on src/Raven.Server/ServerWide/Commands/Sharding/StartBucketMigrationCommand.cs:
```diff
@@ -114,11 +116,30 @@ private void ProcessSubscriptionsForMigration(ClusterOperationContext context, S
+        private void AssertDestinationShardExists(ShardingConfiguration shardingConfiguration)
+        {
+            if (shardingConfiguration.Shards.ContainsKey(DestinationShard) == false)
+                throw new RachisApplyException($"Destination shard {DestinationShard} doesn't exists");
```
if we will get this, there is no database name in the ex, isn't that a problem? same with ex few lines below

aviv86 replied:
fixed

ppekrol commented on src/Raven.Server/ServerWide/Commands/Sharding/DeletePrefixedShardingSettingCommand.cs:
```diff
@@ -0,0 +1,58 @@
+        public override void UpdateDatabaseRecord(DatabaseRecord record, long etag)
+        {
+            int prefixIndex = record.Sharding.Prefixed.BinarySearch(Prefix, PrefixedSettingComparer.Instance);
+            Debug.Assert(prefixIndex >= 0, $"prefix {Prefix.Prefix} doesn't exist");
```
in release index ouf of bounds if value is not found?

aviv86 replied:
fixed

ayende commented on src/Raven.Server/Documents/Sharding/Handlers/Admin/ShardedAdminShardingHandler.cs:
```diff
@@ -13,5 +27,177 @@ public async Task ExecuteMoveDocuments()
+                var clusterTopology = ServerStore.GetClusterTopology(context);
+                var urls = shardingConfiguration.Orchestrator.Topology.Members.Select(clusterTopology.GetUrlFromTag).ToArray();
+
+                if (await AssertNoDocumentsStartingWith(context, setting.Prefix, urls) == false)
+                    throw new InvalidOperationException(
+                        $"Cannot remove prefix '{setting.Prefix}' from {nameof(ShardingConfiguration)}.{nameof(ShardingConfiguration.Prefixed)}. " +
+                        $"There are existing documents in database '{DatabaseName}' that start with '{setting.Prefix}'. " +
+                        "In order to remove a sharding by prefix setting, you cannot have any documents in the database that starts with this prefix.");
+
+                var cmd = new DeletePrefixedShardingSettingCommand(setting, DatabaseName, GetRaftRequestIdFromQuery());
+                var (raftIndex, _) = await ServerStore.SendToLeaderAsync(cmd);
+
+                await DatabaseContext.ServerStore.WaitForExecutionOnRelevantNodesAsync(context, shardingConfiguration.Orchestrator.Topology.Members, raftIndex);
```
More holding a read transaction during remote calls

aviv86 replied:
fixed

ayende commented on src/Raven.Server/Documents/Sharding/Handlers/Admin/ShardedAdminShardingHandler.cs:
```diff
@@ -13,5 +27,177 @@ public async Task ExecuteMoveDocuments()
+            using (ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            using (context.OpenReadTransaction())
+            {
+                var json = await context.ReadForMemoryAsync(RequestBodyStream(), GetType().Name);
```
Here too, for that matter. Can move that higher before the transaction starts

aviv86 replied:
fixed

karmeli87 commented on src/Raven.Server/ServerWide/Commands/Sharding/UpdatePrefixedShardingSettingCommand.cs:
```diff
@@ -0,0 +1,37 @@
+
+        public override void UpdateDatabaseRecord(DatabaseRecord record, long etag)
+        {
+            var location = record.Sharding.Prefixed.BinarySearch(Setting, PrefixedSettingComparer.Instance);
```
can you call `UpdatePrefixedShardingSettingCommand` without having a prior `Prefixed` set? this will cause NRE..

karmeli87 reviewed (approved):
Well done! looks great :-)

arekpalinski commented:
Status: this is waiting for studio work

ml054 commented:
Can we also rebase as I can see we have conflicts here?

aviv86 commented:
> Can we also rebase as I can see we have conflicts here?

done

ppekrol commented on src/Raven.Server/Documents/Handlers/Admin/AdminShardingHandler.cs:
```diff
@@ -22,6 +23,24 @@ public async Task ExecuteMoveDocuments()
             HttpContext.Response.StatusCode = (int)HttpStatusCode.NoContent;
         }
 
+        [RavenAction("/databases/*/admin/sharding/prefixed/add", "POST", AuthorizationStatus.DatabaseAdmin)]
```
this is REST, so remove `/add` and switch to `PUT` ?

ppekrol commented on src/Raven.Server/Documents/Handlers/Admin/AdminShardingHandler.cs:
```diff
@@ -22,6 +23,24 @@ public async Task ExecuteMoveDocuments()
+            throw new NotSupportedInShardingException("This operation is not available from a specific shard");
+        }
+
+        [RavenAction("/databases/*/admin/sharding/prefixed/delete", "DELETE", AuthorizationStatus.DatabaseAdmin)]
```
remove `/delete` ?

ppekrol commented on src/Raven.Server/Documents/Handlers/Admin/AdminShardingHandler.cs:
```diff
@@ -22,6 +23,24 @@ public async Task ExecuteMoveDocuments()
+            throw new NotSupportedInShardingException("This operation is not available from a specific shard");
+        }
+
+        [RavenAction("/databases/*/admin/sharding/prefixed/update", "POST", AuthorizationStatus.DatabaseAdmin)]
```
remove `/update` ?

ppekrol commented on src/Raven.Client/ServerWide/Sharding/AddPrefixedShardingSettingOperation.cs:
```diff
@@ -0,0 +1,31 @@
+
+        public AddPrefixedShardingSettingOperation(PrefixedShardingSetting setting)
+        {
+            _setting = setting;
```
assert at least for null

ppekrol commented on src/Raven.Client/ServerWide/Sharding/DeletePrefixedShardingSettingOperation.cs:
```diff
@@ -0,0 +1,34 @@
+
+        public DeletePrefixedShardingSettingOperation(string prefix)
+        {
+            _prefix = prefix;
```
assert for null or empty at least?

ppekrol commented on src/Raven.Server/Documents/Sharding/Handlers/Admin/ShardedAdminShardingHandler.cs:
```diff
@@ -13,5 +29,182 @@ public async Task ExecuteMoveDocuments()
+                var exists = shardingConfiguration.Prefixed.BinarySearch(setting, PrefixedSettingComparer.Instance) >= 0;
+                if (exists)
+                    throw new InvalidOperationException(
+                        $"Prefix '{setting.Prefix}' already exists in {nameof(ShardingConfiguration)}.{nameof(ShardingConfiguration.Prefixed)}. please use '{nameof(UpdatePrefixedShardingSettingOperation)} operation'");
```
please => Please
add `.` (dot`) at the end.
there is a `'` there at the end as well?

ppekrol commented on src/Raven.Server/ServerWide/ClusterCommandsVersionManager.cs:
```diff
@@ -181,9 +181,14 @@ public sealed class ClusterCommandsVersionManager
             [nameof(EditDataArchivalCommand)] = 60_000,
+
             [nameof(UpdateResponsibleNodeForTasksCommand)] = UpdateResponsibleNodeForTasksCommand.CommandVersion,
+            [nameof(AddPrefixedShardingSettingCommand)] = 60_001,
```
61_000

ppekrol commented on src/Raven.Server/ServerWide/ClusterCommandsVersionManager.cs:
```diff
@@ -181,9 +181,14 @@ public sealed class ClusterCommandsVersionManager
+
             [nameof(UpdateResponsibleNodeForTasksCommand)] = UpdateResponsibleNodeForTasksCommand.CommandVersion,
+            [nameof(AddPrefixedShardingSettingCommand)] = 60_001,
+            [nameof(DeletePrefixedShardingSettingCommand)] = 60_001,
```
61_000

ppekrol commented on src/Raven.Server/ServerWide/ClusterCommandsVersionManager.cs:
```diff
@@ -181,9 +181,14 @@ public sealed class ClusterCommandsVersionManager
             [nameof(UpdateResponsibleNodeForTasksCommand)] = UpdateResponsibleNodeForTasksCommand.CommandVersion,
+            [nameof(AddPrefixedShardingSettingCommand)] = 60_001,
+            [nameof(DeletePrefixedShardingSettingCommand)] = 60_001,
+            [nameof(UpdatePrefixedShardingSettingCommand)] = 60_001
```
61_000

ppekrol reviewed (approved):
LGTM, beside few small remarks that I've left.

aviv86 commented:
> LGTM, beside few small remarks that I've left.

done
cc @ml054

ml054 commented:
@aviv86 Can you rebase? 
@aviv86 Is there anything left in studio, that you are aware of?

arekpalinski commented:
@aviv86 can you please rebase against 6.1 branch?

Commit:
* RavenDB-17793 : Prefixed Sharding

RavenDB-17793 : address PR comments

RavenDB-17793 Add sharding prefixes step to database creator

RavenDB-17793 : fixes after rebase

RavenDB-17793 : address PR comments - throw instead of debug asserts, add more info to error messages

RavenDB-17793 : add tests cases for (1) querying with shard context (2) prefixes should get precedence over anchoring when choosing bucket for id

RavenDB-17793 : address PR comments - avoid doing remote calls within a read tx

RavenDB-17793 : fix failing tests

RavenDB-17793 : minor refactoring + cleanup

RavenDB-17793 : remove BlockPrefixedSharding from ShardingStore

RavenDB-17793 : fix bug in DeletePrefixedSettingCommand - need to remove all associated bucket ranges for this prefix (do not rely on number of shards in prefix setting)

RavenDB-17793 : add more test scenarios

RavenDB-17793 : do not allow to add prefix setting with no shards

RavenDB-17793 : pass prefix to StartBucketMigrationCommand so that we can find the prefix setting by name (sorted) instead of searching by bucket range start (needs to be ordered first)

RavenDB-17793 : when adding a new prefix, reuse old bucket ranges (i.e. bucket ranges that were given to a cretin prefix which was removed) if possible

RavenDB-17793 : do not allow to remove shard from database if this shard is part of any prefix settings

RavenDB-17793 : update prefix setting - don't allow to remove shard if there are bucket ranges mapped to this shards. update command should not modify BucketRanges at all

RavenDB-17793 : throw on attempt to move bucket from prefixed range if destination shard is not a part of shards list in the prefix setting

RavenDB-17793 : use binary search for lookups in prefixed list

RavenDB-17793 : fix bug in sharded restore - do not fill PrefixedSharding on AddDatabaseCommand with no RaftCommandIndex when coming from restore (in restore we take the prefixed setting as is from smuggler record)

RavenDB-17793 : prefixed settings list should be sorted in a descending manner, in order to support having a prefix of prefix

RavenDB-17793 : ensure that prefixes settings cannot be modified via database record. add dedicated operation for updating prefix configuration

aviv86 commented:
> 

squashed everything into a single commit and rebased against 6.1

aviv86 commented:
> @aviv86 can you please rebase against 6.1 branch?


squashed everything into a single commit and rebased against 6.1

aviv86 commented:
> @aviv86 Can you rebase? @aviv86 Is there anything left in studio, that you are aware of?

I changed the `/prefixed` endpoints urls in `ShardedAdminShardingHandler` (https://github.com/ravendb/ravendb/pull/18098#discussion_r1660776087)

ppekrol commented on src/Raven.Server/Documents/Handlers/Admin/AdminShardingHandler.cs:
```diff
@@ -22,6 +23,24 @@ public async Task ExecuteMoveDocuments()
+        [RavenAction("/databases/*/admin/sharding/prefixed", "PUT", AuthorizationStatus.DatabaseAdmin)]
+        public Task AddPrefixedShardingSetting()
+        {
+            throw new NotSupportedInShardingException("This operation is not available from a specific shard");
```
this is not a correct exception? regular NotSupportedException should be here?

ppekrol commented on src/Raven.Server/Documents/Handlers/Admin/AdminShardingHandler.cs:
```diff
@@ -22,6 +23,24 @@ public async Task ExecuteMoveDocuments()
+        [RavenAction("/databases/*/admin/sharding/prefixed", "DELETE", AuthorizationStatus.DatabaseAdmin)]
+        public Task DeletePrefixedShardingSetting()
+        {
+            throw new NotSupportedInShardingException("This operation is not available from a specific shard");
```
this is not a correct exception? regular NotSupportedException should be here?

ppekrol commented on src/Raven.Server/Documents/Handlers/Admin/AdminShardingHandler.cs:
```diff
@@ -22,6 +23,24 @@ public async Task ExecuteMoveDocuments()
+        [RavenAction("/databases/*/admin/sharding/prefixed", "POST", AuthorizationStatus.DatabaseAdmin)]
+        public Task UpdatePrefixedShardingSetting()
+        {
+            throw new NotSupportedInShardingException("This operation is not available from a specific shard");
```
this is not a correct exception? regular NotSupportedException should be here?

ppekrol commented on src/Raven.Server/Documents/Sharding/Handlers/Admin/ShardedAdminShardingHandler.cs:
```diff
@@ -13,5 +29,182 @@ public async Task ExecuteMoveDocuments()
+
+                await DatabaseContext.ServerStore.WaitForExecutionOnRelevantNodesAsync(context, shardingConfiguration.Orchestrator.Topology.Members, raftIndex);
+
+                HttpContext.Response.StatusCode = (int)HttpStatusCode.NoContent;
```
use NoContent() or NoContentStatus() method

ppekrol commented on src/Raven.Server/Documents/Sharding/Handlers/Admin/ShardedAdminShardingHandler.cs:
```diff
@@ -13,5 +29,182 @@ public async Task ExecuteMoveDocuments()
+
+                await DatabaseContext.ServerStore.WaitForExecutionOnRelevantNodesAsync(context, shardingConfiguration.Orchestrator.Topology.Members, raftIndex);
+
+                HttpContext.Response.StatusCode = (int)HttpStatusCode.NoContent;
```
use NoContent() or NoContentStatus() method

ppekrol commented on src/Raven.Server/Documents/Sharding/Handlers/Admin/ShardedAdminShardingHandler.cs:
```diff
@@ -13,5 +29,182 @@ public async Task ExecuteMoveDocuments()
+
+                await DatabaseContext.ServerStore.WaitForExecutionOnRelevantNodesAsync(context, shardingConfiguration.Orchestrator.Topology.Members, raftIndex);
+
+                HttpContext.Response.StatusCode = (int)HttpStatusCode.NoContent;
```
use NoContent() or NoContentStatus() method

ppekrol commented on src/Raven.Server/Documents/Sharding/Handlers/Admin/ShardedAdminShardingHandler.cs:
```diff
@@ -13,5 +29,182 @@ public async Task ExecuteMoveDocuments()
+            }
+        }
+
+        private async Task<bool> AssertNoDocumentsStartingWith(JsonOperationContext context, string prefix, string[] urls, string database = null)
```
Async in method name

ppekrol reviewed (approved):
Small remarks

Commit:
RavenDB-17793 : address minor PR comments

aviv86 commented:
> Small remarks

fixed
